torch/distributed/_tensor/dispatch.py

```diff
@@ -149,14 +149,47 @@ def operator_dispatch(
     needs_redistribute = suggested_input_schema is not op_schema
 
     if mesh is not None and mesh.get_coordinate() is None:
-        # if we are on a non-participating device, we simply return
-        # an empty tensor for now.
-        # TODO: what if the op returns a non-tensor value, what if
-        # the op returns a list of tensors, we need to figure out
-        # a consistent way to handle that, and also need to figure
-        # out if we should communicate the result to non-participating
-        # ranks (i.e. a.sum() -> scalar, maybe we should set to 0)
-        local_results = torch.tensor([])
+        # if we are on a non-participating device, we should return
+        # a default value of some type consistent with other
+        # participating devices. This type can be obtained from
+        # the `OutputSharding` object by the following logic:
+        # 0. if output_spec and schema_suggestions are both None, unsuccessful
+        # sharding propagation. This has been handled by assertion on
+        # schema_suggestions
+        # 1. if output_spec is None and schema_suggestions is same as the
+        # input, the return value is a non-tensor (scalar) value.
+        # 2. else, a successful propagation. output_spec contains the return
+        # type (scalar, DTensor, List[DTensor])
+        def _default_ret_value(spec):
+            if spec is None:
+                return torch.tensor(0)
```

Comment:
if `spec` is None, it means that this op is not returning tensor value, it would return other types, i.e. `int`, `bool`, I think we would need to communicate the non-tensor result to all ranks for this case, let's add the TODO you mentioned later to here, any default value here seems not making much sense, we should probably error out now.